### DISCUSSION TITLE
Allure-js-commons /SDK/Reporting, not able to add stopTime to the test

### DISCUSSION THREAD

**Discussion by vanBadAsselt** [allure-framework/allure2#2644] [Ideas & Feedback]
TL;DR: It seems that there is an issue in [allure-js-commons > SDK > reporter runtime > calculate timings](https://github.com/allure-framework/allure-js/blob/main/packages/allure-js-commons/src/sdk/reporter/ReporterRuntime.ts#L565) that prevents the stop time to be added. When stop time is provided, in the allure-results.json it results in `"stop": { "stop": 1721054909078, "duration": 82743.77525 } `, which shows an empty result in Allure

Context: I created an NX plugin to convert K6 test results into Allure results using the allure-js-commons lib (3.0.0-beta.5), this works but so far I'm not able to show the test duration. I have a start and stop time that I convert to a 13 digit epoch time which I then assign to the test result: 

- Start time = 2024-07-15T16:47:07.331066+02:00 = 1721054827331
- Stop time = 2024-07-15T16:48:29.078124+02:00 = 1721054909078

When I add the times through `updateTest` like below, the start time matches but the stoptime is 1721114020529

```
 reporter.updateTest(testUuid!, (testResult: any) => {
      testResult.status = testStatus;
      testResult.stage = Stage.FINISHED;
      testResult.start = testTimeStart;
      testResult.stop = testTimeStop;
  });
```
When I try `stopTest` like below, stop will result in this object, which doesn't upload in Allure (shows empty result), because probably stop should be a number:   `"stop": { "stop": 1721054909078, "duration": 82743.77525 } `

```
 reporter.stopTest(testUuid, {
      stop: testTimeStop,
      duration: testDuration
    });
```

If I look in the code of the [Reporter](https://github.com/allure-framework/allure-js/blob/main/packages/allure-js-commons/src/sdk/reporter/ReporterRuntime.ts), it looks like the issue lies in `calculateTimings`

```
const startStop = this.#calculateTimings(testResult.start, opts?.stop, opts?.duration);

```

My observations:
- It seems that it's only possible to pass the stop time through `opts?.stop`, 
- If this option is passed, calculateTimings returns `const result: { start?: number; stop?: number } = { start, stop };`
- This result is assigned to the stop object which breaks the allure result

I attached the following as a zip:
* K6 test results
* Executor I created + allureEnums needed
* Allure output that is broken (if stop is a number, and the label is a string it works)
[AllureQs.zip](https://github.com/user-attachments/files/16266568/AllureQs.zip)

Suggested fix :
- Fix return of calculateTimings
- Use testResult.stop when its provided

**Comment by baev** [comment]
Sorry, I don't quite understand the issue. The way `calculateTimings` works:

1. You must provide either `stop` or `duration` for the `stopTest` method. The step is expected to be the Unix timestamp in milliseconds and duration — the positive integer representing test duration in milliseconds.
2. If no `stop` or `duration` is provided, `ReporterRuntime` will use the current timestamp.
3. If both `stop` and `duration` are provided, `ReporterRuntime` will recalculate and update the start time (as `start = stop - duration`).
4. If only `stop` is provided, it is set to test the result.
5. If only `duration` is provided, `ReporterRuntime` will calculate `stop` from `start` (if it's provided) as `stop = start + duration`).
6.  If only `duration` is provided and no `start` is set in the result — `ReporterRuntime` will consider `stop` to be `Date.now()` and calculate `start` as `start = stop - duration`.

BTW, if you're converting one JSON result to another, there is no need to use `ReporterRuntime` — you can create a `TestResult` object and write it using `FileSystemWriter`.

**Reply by vanBadAsselt** [reply]
thanks for your quick reply @baev , what I mean in regards to `calculateTimings`  :
1. If I provide `stop` in Unix timestamp in ms, the JSON generated for Allure contains this `stop` object: 
 `"stop": { "stop": 1721054909078, "duration": 82743.77525 } `, If I try to upload this in Allure, I'm seeing an empty launch (if I upload a result json with `"stop": 1721054909078`, it does work. 
2. I'm using the FileSystemWriter according to the advice given here: https://github.com/orgs/allure-framework/discussions/2610, bc it seemed quite relatable to my use case (my impl can be found in the executor.ts in the zip file I provided). If there is a better way to do this, I'd love to hear it and get an example.

**Reply by baev** [reply]
1. There is a test for that. See https://github.com/allure-framework/allure-js/blob/v3.0.0-beta.7/packages/allure-js-commons/test/sdk/reporter/ReporterRuntime.spec.ts#L42-L55

**Reply by baev** [reply]
2. You can create `TestResult` and write it using `FileSystemWriter`. The `ReporterRuntime` is the utility class that creates test results on the fly based on real-time events.

``` ts
const testResult: TestResult = {
    uuid: randomUUID(),
    name: "a test name",
    historyId: md5("something"),
    status: Status.PASSED,
    statusDetails: {},
    stage: Stage.FINISHED,
    steps: [],
    attachments: [],
    parameters: [],
    labels: [],
    links: [],
  }

const writer = new FileSystemWriter({ resultsDir: "./out/allure-results"});
writer.writeResult(testResult);
```
